torch/_subclasses/meta_utils.py

```diff
@@ -89,19 +118,274 @@ def is_sparse_coo(t):
+
+        # TODO: Is it important to enable torch.inference_mode before querying
+        # these values?
+        return MetaTensorDesc(
+            id=self.get_tensor_id(t),
+            storage=storage,
+            is_inference=t.is_inference(),
+            is_leaf=is_leaf,
+            requires_grad=t.requires_grad,
+            # NB: ndim should be OK too but there is a disaster at
+            # python test/dynamo/test_subclasses.py -k test_user_overidden_property_unsupported
+            # Actually, this means that we have a little bit of a problem
+            # here, which is that there is some sensitivity to how exactly an
+            # access is done if you have a __torch_function__ subclass.  Maybe
+            # should disable torch function before doing accesses?
+            ndim=t.dim(),
+            dtype=t.dtype,
+            is_sparse=is_sparse,
+            is_mkldnn=is_mkldnn,
+            is_functorch_wrapped=is_functorch_wrapped,
+            is_batchedtensor=is_batchedtensor_v,
+            is_gradtrackingtensor=is_gradtrackingtensor_v,
+            is_view=is_view,
+            is_conj=t.is_conj(),
+            is_neg=t.is_neg(),
+            is_traceable_wrapper_subclass=is_traceable_wrapper_subclass_v,
+            is_nested=is_nested,
+            layout=layout,
+            size=t.size(),
+            stride=stride,
+            storage_offset=storage_offset,
+            dynamo_dynamic_indices=list(getattr(t, "_dynamo_dynamic_indices", set())),
+            sparse_dim=t.sparse_dim()
+            if t.is_sparse or is_sparse_compressed(t)
+            else None,
+            dense_dim=t.dense_dim() if t.is_sparse or is_sparse_compressed(t) else None,
+            is_coalesced=t.is_coalesced() if t.is_sparse else None,
+            # TODO: I actually think recursing here is correct, but we have at
+            # least an infinite cycle from base -> values -> base
+            # https://github.com/pytorch/pytorch/issues/122089
+            crow_indices=self.describe_tensor(t.crow_indices(), recurse=False)
+            if recurse and t.layout in {torch.sparse_csr, torch.sparse_bsr}
+            else None,
+            col_indices=self.describe_tensor(t.col_indices(), recurse=False)
+            if recurse and t.layout in {torch.sparse_csr, torch.sparse_bsr}
+            else None,
+            ccol_indices=self.describe_tensor(t.ccol_indices(), recurse=False)
+            if recurse and t.layout in {torch.sparse_csc, torch.sparse_bsc}
+            else None,
+            row_indices=self.describe_tensor(t.row_indices(), recurse=False)
+            if recurse and t.layout in {torch.sparse_csc, torch.sparse_bsc}
+            else None,
+            values=self.describe_tensor(t.values(), recurse=False)
+            if recurse and is_sparse_compressed(t)
+            else None,
+            grad=self.describe_tensor(safe_grad(t))
+            if safe_grad(t) is not None
+            else None,
```

Comment:
Many of these fields are mutually exclusive. It might be easier to parse all of this is we had:

SparseMeta vs BatchedMeta vs StridedTensor meta or something of the like that encapsulates the subfields

Reply:
I actually originally set it up this way, but because I want this to be a stable serialization format, it's more convenient to have all the subfields flattened in one struct, because JSON is bad at unions and unions are also bad for BC.